edgelet/edgelet-http/src/lib.rs

```diff
@@ -69,6 +77,103 @@ const TCP_SCHEME: &str = "tcp";
 #[cfg(target_os = "linux")]
 const FD_SCHEME: &str = "fd";
 
+#[derive(Clone)]
+pub struct PemCertificate {
+    cert: Vec<u8>,
+    key: Option<Vec<u8>>,
+    username: Option<String>,
+    password: Option<String>,
+}
+
+impl PemCertificate {
+    pub fn new(
+        cert: Vec<u8>,
+        key: Option<Vec<u8>>,
+        username: Option<String>,
+        password: Option<String>,
+    ) -> Self {
+        PemCertificate {
+            cert,
+            key,
+            username,
+            password,
+        }
+    }
+
+    pub fn get_certificate(&self) -> &[u8] {
+        &self.cert
+    }
+
+    pub fn from<C: Certificate>(id_cert: &C) -> Result<Self, Error> {
+        let cert = id_cert
+            .pem()
+            .map(|cert_buffer| cert_buffer.as_ref().to_owned())
+            .context(ErrorKind::IdentityCertificate)?;
+
+        let key = match id_cert.get_private_key() {
+            Ok(Some(PrivateKey::Ref(ref_))) => Some(ref_.into_bytes().clone()),
+            Ok(Some(PrivateKey::Key(KeyBytes::Pem(buffer)))) => Some(buffer.as_ref().to_vec()),
+            _ => None,
+        };
+
+        Ok(PemCertificate::new(cert, key, None, None))
+    }
+
+    pub fn get_identity(&self) -> Result<Identity, Error> {
+        let mut certs =
+            X509::stack_from_pem(&self.cert).with_context(|_| ErrorKind::IdentityCertificate)?;
+
+        // the first cert is the identity cert and the other certs are part of the CA
+        // chain; we skip the server cert and build an OpenSSL cert stack with the
+        // other certs
+        let mut ca_certs = Stack::new().with_context(|_| ErrorKind::IdentityCertificate)?;
+        for cert in certs.split_off(1) {
+            ca_certs
+                .push(cert)
+                .with_context(|_| ErrorKind::IdentityCertificate)?;
+        }
+
+        let key = match &self.key {
+            Some(k) => PKey::private_key_from_pem(&k),
+            None => return Err(Error::from(ErrorKind::IdentityPrivateKey)),
+        }?;
+
+        let identity_cert = &certs[0];
+
+        let mut builder = Pkcs12::builder();
+        builder.ca(ca_certs);
+        let pkcs_certs = builder
```

Comment:
Since this is a straightforward use of the builder, you can just inline it into `let pkcs_certs = Pkcs12::builder().ca().build()` instead of splitting off the builder into its own binding.

Reply:
This unfortunately does not work since the API does not lend itself to this sort of usage.

Reply:
Right, I missed that `Pkcs12Builder::build` consumes `self`